### PR TITLE
Add support for form submissions

### PR DESCRIPTION
**Pull request by RaeesBhatti**
**- Summary**
Add support for form submissions.
Fixes: https://github.com/netlify/cli/issues/447

**- Test plan**
* Run `netlify dev`
* Make a POST request with url encoded data

**- Description for the changelog**
Transform the body of POST requests with Form data to match a `submission-created` event in Production.

**- A picture of a cute animal (not mandatory but encouraged)**
🐰

### CODE REVIEW

File: src/commands/dev/index.js [hunk 1]
```diff
@@ -172,7 +175,25 @@ async function startProxy(settings, addonUrls, configPath, projectDir) {
     configPath,
   })
 
+  mkdirp.sync(path.resolve(projectDir, '.netlify'))
```
Comment (ehmicky) [inline src/commands/dev/index.js]: Would `await util.promisify(mkdirp)(path.resolve(projectDir, '.netlify'))` work instead?

File: src/utils/forms.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+const FORM_URLENCODED = 'application/x-www-form-urlencoded'
+
+function collectRequestBody(request, callback) {
```
Comment (ehmicky) [inline src/utils/forms.js]: I think you might want to use a library to read `request` body instead? Libraries like `raw-body` adds lots of tricks and handle some edge cases when doing this.

```js
const getRawBody = require('raw-body')
const body = String(await getRawBody(request))
```

File: src/commands/dev/index.js [hunk 3]
```diff
@@ -172,7 +175,25 @@ async function startProxy(settings, addonUrls, configPath, projectDir) {
...
+  const csvWriter = csvWriterPkg.createArrayCsvWriter({
+    alwaysQuote: true,
+    header: ['timestamp', 'path', 'body'],
+    path: formSubmissionsFile,
+    append: true,
+  })
+
   const server = http.createServer(function(req, res) {
+    if (req.method === 'POST') {
+      collectRequestBody(req, async body => {
```
Comment (ehmicky) [inline src/commands/dev/index.js]: The `Content-Type` response header should be set to `text/csv`?

File: src/commands/dev/index.js [hunk 4]
```diff
@@ -172,7 +175,25 @@ async function startProxy(settings, addonUrls, configPath, projectDir) {
     configPath,
   })
 
+  await mkdirp(path.resolve(projectDir, '.netlify'))
```
Comment (ehmicky) [inline src/commands/dev/index.js]: `mkdirp` requires `util.promisify()` to be used with `await` I believe.

Comment (RaeesBhatti) [reply src/commands/dev/index.js]: The latest version has supported for promises by default

Comment (ehmicky) [reply src/commands/dev/index.js]: Oh right, I missed that you also upgraded `mkdirp` :+1:

**Comment by erquhart** [comment]
@fool correct if I'm wrong but you weren't looking for actual submissions here, just a successful response and trigger the form submission triggered function if it exists. The form submission itself would not be retained. Sound about right?

File: src/commands/dev/index.js (hunk 3 above)
Comment (fool) [reply src/commands/dev/index.js:190]: most folks return only html from their form submission, so probably don't need to get too fancy.

**Comment by fool** [comment]
yes, that was indeed my thinking, @erquhart

File: src/utils/forms.js (hunk 2 above)
Comment (RaeesBhatti) [reply src/utils/forms.js:3]: Oh nice, thanks!

**Comment by RaeesBhatti** [comment]
@fool Do you have any advice on where to look for sample `submission-created ` event payload? I need the event shape to emulate that in `netlify dev`.

**Comment by fool** [comment]
I don't have good examples of the payloads, aside from what is in the public doc:

https://docs.netlify.com/functions/trigger-on-events/#payload

**Comment by fool** [comment]
Would be good to get a "dump the input" function example spun up.  I bet @depadiernos could write one in about 5 minutes, but it's not something I can just toss out since I'm no javascript pro

**Pull request merged**